src/protocols/secure_channel/PASESession.cpp

```diff
@@ -346,133 +349,200 @@ CHIP_ERROR PASESession::DeriveSecureSession(SecureSession & session, SecureSessi
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    uint16_t responderSessionId = 0;
+    uint8_t random[kPBKDFParamRandomNumberSize];
 
-    // This the fixed part of the message. The variable part of the message contains the salt.
-    // The length of the variable part is determined by the salt length in the fixed header.
-    static_assert(CHAR_BIT == 8, "Assuming that sizeof returns octets");
-    size_t fixed_resplen = kPBKDFParamRandomNumberSize + sizeof(uint64_t) + sizeof(uint32_t);
+    uint32_t decodeTagIdSeq = 0;
+    uint32_t iterCount      = 0;
+    uint32_t saltLength     = 0;
+    const uint8_t * salt;
 
     ChipLogDetail(SecureChannel, "Received PBKDF param response");
 
-    VerifyOrExit(resp != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
-    VerifyOrExit(resplen >= fixed_resplen, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    SuccessOrExit(err = mCommissioningHash.AddData(ByteSpan{ msg->Start(), msg->DataLength() }));
 
-    {
-        // Let's skip the random number portion of the message
-        const uint8_t * msgptr = &resp[kPBKDFParamRandomNumberSize];
-        uint64_t iterCount     = chip::Encoding::LittleEndian::Read64(msgptr);
-        uint32_t saltlen       = chip::Encoding::LittleEndian::Read32(msgptr);
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    // Initiator's random value
+    SuccessOrExit(err = tlvReader.GetBytes(random, sizeof(random)));
+
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    // Responder's random value
+    SuccessOrExit(err = tlvReader.GetBytes(random, sizeof(random)));
 
-        VerifyOrExit(resplen == fixed_resplen + saltlen, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
+    SuccessOrExit(err = tlvReader.Get(responderSessionId));
 
-        // Specifications allow message to carry a uint64_t sized iteration count. Current APIs are
-        // limiting it to uint32_t. Let's make sure it'll fit the size limit.
-        VerifyOrExit(CanCastTo<uint32_t>(iterCount), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    ChipLogDetail(SecureChannel, "Peer assigned session key ID %d", responderSessionId);
+    SetPeerKeyId(responderSessionId);
 
-        // Update commissioning hash with the received pbkdf2 param response
-        err = mCommissioningHash.AddData(ByteSpan{ resp, resplen });
+    if (mHavePBKDFParameters)
+    {
+        err = SetupSpake2p(iterCount, ByteSpan(mSalt, mSaltLength));
```

Comment:
Why does this use iterCount (set to zero at the top and never otherwise touched) rather than mIterationCount? How does that work? I think this path is not tested

Reply:
Yes, it should be `mIterationCount`. Thanks for catching it.
Need a test for this. I'll file an issue to track it.

Reply:
Yes, we need a test of some more conditions in PASE and CASE, especially regarding bad message contents.

Reply:
Added a TODO for the pending test.